# Описание задачи

#### Необходимо обработать сырые данные поступившие от парсера

# 1. Подключение библиотек

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import folium
from folium import Marker, Map
from folium.plugins import MarkerCluster

pd.set_option('display.max_columns', None)

# 2. Загрузка данных

In [2]:
data1 = pd.read_excel(r'C:\Users\User\Downloads\Тестовое задание 2\hotels_parsing_result.xlsx', sheet_name='platform1')

In [3]:
data2 = pd.read_excel(r'C:\Users\User\Downloads\Тестовое задание 2\hotels_parsing_result.xlsx', sheet_name='platfrom2')

# 3. Открытие данных

In [4]:
data1.head()

id                 create_time                             title  \
0  81842  2023-09-26 14:45:42.910962  Апартаменты Baltia Западный Пляж   
1  86341  2023-09-26 15:01:55.184815                     Отель Авиатор   
2  86420  2023-09-26 15:02:12.005085                         Клеопатра   
3  86421  2023-09-26 15:02:12.225334                           Татьяна   
4  85898  2023-09-26 15:00:19.217359                Мини-Отель Чемодан   

  hotel_type_original          city                                address  \
0           Apartment  Зеленоградск  улица Приморская, д. 31, Зеленоградск   
1               Hotel           Уфа             улица Мушникова, д.28, Уфа   
2               Hotel           Уфа       Коммунистическая ул., д. 53, Уфа   
3               Hotel         Сибай            Учалинская улица, 37, Сибай   
4          Mini-hotel   Стерлитамак      улица Дружбы, д.28 Д, Стерлитамак   

   rating  rating_5  review_count  region  star_rating  rooms_count  \
0    10.0      5.00             0       2          NaN          NaN   
1     8.7      4.35             5       2          NaN          8.0   
2     6.3      3.15            10       2          4.0         15.0   
3    10.0      5.00             2       2          NaN         15.0   
4     8.0      4.00             0       2          2.0         23.0   

   contact_social  description  email  phone  website       uid  \
0             NaN          NaN    NaN    NaN      NaN  10595357   
1             NaN          NaN    NaN    NaN      NaN   8332081   
2             NaN          NaN    NaN    NaN      NaN   7724907   
3             NaN          NaN    NaN    NaN      NaN   6598197   
4             NaN          NaN    NaN    NaN      NaN   8622634   

                 parsing_time        lat        lon  
0  2023-09-26 20:56:02.013106  54.951103  20.456526  
1  2023-09-26 20:56:02.013106  54.783913  56.116150  
2  2023-09-26 20:56:02.013106  54.726463  55.946445  
3  2023-09-26 20:56:02.013106  52.719230  58.658596  
4  2023-09-26 20:56:02.013106  53.642567  55.932686

In [5]:
data2.head()

id                 create_time                  title  \
0   4608  2023-09-14 15:46:07.855977         Уют, гостиница   
1   4815  2023-09-14 15:46:48.332492      Гостиница Ашкадар   
2   4750  2023-09-14 15:46:35.659931  Островок, база отдыха   
3   4786  2023-09-14 15:46:42.672665     Ардан, база отдыха   
4  16310  2023-09-15 11:56:48.785266              Гостиница   

  hotel_type_original         city  \
0               hotel          NaN   
1                 NaN  sterlitamak   
2          hotel_rest          NaN   
3          hotel_rest          NaN   
4               hotel    beloretsk   

                                             address  rating  rating_5  \
0  Россия, Республика Башкортостан, Аскинский рай...     NaN       NaN   
1  Россия, Республика Башкортостан, Стерлитамак г...     NaN       NaN   
2  Россия, Республика Башкортостан, Гафурийский р...     NaN       NaN   
3  Россия, Республика Башкортостан, Баймакский ра...     NaN       NaN   
4  Россия, Республика Башкортостан, Белорецкий ра...     NaN       NaN   

   review_count  region  star_rating  rooms_count contact_social description  \
0           NaN       2          NaN          5.0            NaN         NaN   
1           NaN       2          NaN          NaN            NaN         NaN   
2           NaN       2          NaN          NaN            NaN         NaN   
3           NaN       2          NaN          NaN            NaN         NaN   
4           NaN       2          NaN          NaN            NaN         NaN   

  email phone website                uid                parsing_time  \
0   NaN   NaN     NaN  70000001064499291  2023-09-14 22:44:57.013238   
1   NaN   NaN     NaN   7600610350202940  2023-09-14 22:44:57.054250   
2   NaN   NaN     NaN  70000001076845858  2023-09-14 22:44:57.043234   
3   NaN   NaN     NaN  70000001075946655  2023-09-14 22:44:57.049244   
4   NaN   NaN     NaN  70000001031030911  2023-09-15 18:55:23.191778   

         lat        lon  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4  53.969912  58.400557

# 4. Предобработка данных data1

In [6]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5522 entries, 0 to 5521
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   5522 non-null   int64  
 1   create_time          5522 non-null   object 
 2   title                5522 non-null   object 
 3   hotel_type_original  5522 non-null   object 
 4   city                 5522 non-null   object 
 5   address              5522 non-null   object 
 6   rating               5522 non-null   float64
 7   rating_5             5522 non-null   float64
 8   review_count         5522 non-null   int64  
 9   region               5522 non-null   int64  
 10  star_rating          311 non-null    float64
 11  rooms_count          856 non-null    float64
 12  contact_social       0 non-null      float64
 13  description          0 non-null      float64
 14  email                0 non-null      float64
 15  phone                0 non-null      f

#### Массив данных содержит 21 колонку и 5522 строки, из них 5 колонок с контактами, описанием, сайтом, почтой полностью пустые, из чего можно сделать вывод, что данные колонки можно удалить

## 4.1. Пропуски. Удаление пустых колонок

In [7]:
data1.drop('contact_social', axis=1, inplace=True)
data1.drop('description', axis=1, inplace=True)
data1.drop('email', axis=1, inplace=True)
data1.drop('phone', axis=1, inplace=True)
data1.drop('website', axis=1, inplace=True)

In [8]:
data1.head()

id                 create_time                             title  \
0  81842  2023-09-26 14:45:42.910962  Апартаменты Baltia Западный Пляж   
1  86341  2023-09-26 15:01:55.184815                     Отель Авиатор   
2  86420  2023-09-26 15:02:12.005085                         Клеопатра   
3  86421  2023-09-26 15:02:12.225334                           Татьяна   
4  85898  2023-09-26 15:00:19.217359                Мини-Отель Чемодан   

  hotel_type_original          city                                address  \
0           Apartment  Зеленоградск  улица Приморская, д. 31, Зеленоградск   
1               Hotel           Уфа             улица Мушникова, д.28, Уфа   
2               Hotel           Уфа       Коммунистическая ул., д. 53, Уфа   
3               Hotel         Сибай            Учалинская улица, 37, Сибай   
4          Mini-hotel   Стерлитамак      улица Дружбы, д.28 Д, Стерлитамак   

   rating  rating_5  review_count  region  star_rating  rooms_count       uid  \
0    10.0      5.00             0       2          NaN          NaN  10595357   
1     8.7      4.35             5       2          NaN          8.0   8332081   
2     6.3      3.15            10       2          4.0         15.0   7724907   
3    10.0      5.00             2       2          NaN         15.0   6598197   
4     8.0      4.00             0       2          2.0         23.0   8622634   

                 parsing_time        lat        lon  
0  2023-09-26 20:56:02.013106  54.951103  20.456526  
1  2023-09-26 20:56:02.013106  54.783913  56.116150  
2  2023-09-26 20:56:02.013106  54.726463  55.946445  
3  2023-09-26 20:56:02.013106  52.719230  58.658596  
4  2023-09-26 20:56:02.013106  53.642567  55.932686

#### Вывод: колонки с пустыми данными удалены

#### Заменить пропуски в колонке 'star_rating' не представляется возможным, так как данные о гостиницах реальны и замена пустых значений звёздного рейтинга средним значением или медианой не будет отражать реальные рейтинги гостиниц.
#### Такая же ситуация с колонкой 'rooms_count' о количестве номеров в гостинице.
#### ВЫВОД: оставим пропуски как есть

## 4.2. Типы данных

In [9]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5522 entries, 0 to 5521
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   5522 non-null   int64  
 1   create_time          5522 non-null   object 
 2   title                5522 non-null   object 
 3   hotel_type_original  5522 non-null   object 
 4   city                 5522 non-null   object 
 5   address              5522 non-null   object 
 6   rating               5522 non-null   float64
 7   rating_5             5522 non-null   float64
 8   review_count         5522 non-null   int64  
 9   region               5522 non-null   int64  
 10  star_rating          311 non-null    float64
 11  rooms_count          856 non-null    float64
 12  uid                  5522 non-null   int64  
 13  parsing_time         5522 non-null   object 
 14  lat                  5522 non-null   float64
 15  lon                  5522 non-null   f

#### Содержание всех колонок, кроме 'create_time' и 'parsing_time' соответствует типу данных в колонке. Исправим тип данных в колонках о времени

In [10]:
data1['create_time'] = pd.to_datetime(data1['create_time'])

In [11]:
data1['parsing_time'] = pd.to_datetime(data1['parsing_time'])

In [12]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5522 entries, 0 to 5521
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   5522 non-null   int64         
 1   create_time          5522 non-null   datetime64[ns]
 2   title                5522 non-null   object        
 3   hotel_type_original  5522 non-null   object        
 4   city                 5522 non-null   object        
 5   address              5522 non-null   object        
 6   rating               5522 non-null   float64       
 7   rating_5             5522 non-null   float64       
 8   review_count         5522 non-null   int64         
 9   region               5522 non-null   int64         
 10  star_rating          311 non-null    float64       
 11  rooms_count          856 non-null    float64       
 12  uid                  5522 non-null   int64         
 13  parsing_time         5522 non-nul

#### ВЫВОД: тип данных в колонках со временем исправлен

## 4.3. Дубликаты

In [13]:
data1.duplicated().sum()

0

#### ВЫВОД: явных дубликатов не обнаружено

# 5. Исследовательский анализ данных

In [14]:
title = data1['title'].value_counts().reset_index()
title = title.query('count > 1')
title

title  count
0             Апартаменты Infinity of Days      9
1    Апартаменты Apart-Deluxe Легенды Моря      8
2                       Апартаменты у Моря      6
3          Апартаменты у Королевских Ворот      5
4        Апартаменты в центре Калининграда      5
..                                     ...    ...
170                 Апартаменты на Майском      2
171                     Дом отдыха Березки      2
172               Апартаменты Amber lounge      2
173              Апартаменты на Телецентре      2
174                Апартаменты Гагарина 2А      2

[175 rows x 2 columns]

## 5.1. Актуальные данные по гостиницам

#### Финальный список гостиниц, в котором по каждой гостинице - только одна запись с самыми актуальными и полными данными

In [15]:
agg_data1 = data1.groupby('title')['create_time'].max().reset_index()

res_data1 = pd.merge(data1, agg_data1, on=['title', 'create_time'], how='inner')
res_data1

id                create_time                             title  \
0     81842 2023-09-26 14:45:42.910962  Апартаменты Baltia Западный Пляж   
1     86341 2023-09-26 15:01:55.184815                     Отель Авиатор   
2     86420 2023-09-26 15:02:12.005085                         Клеопатра   
3     86421 2023-09-26 15:02:12.225334                           Татьяна   
4     85898 2023-09-26 15:00:19.217359                Мини-Отель Чемодан   
...     ...                        ...                               ...   
5253  86744 2023-09-26 15:03:21.089955                   Podushka Hostel   
5254  86790 2023-09-26 15:03:30.721937                Lux apartament UFA   
5255  87120 2023-09-26 15:04:39.726907                           Биатлон   
5256  87165 2023-09-26 15:04:49.051042                              Лидо   
5257  87190 2023-09-26 15:04:54.242964                 Airport Hotel Ufa   

     hotel_type_original          city  \
0              Apartment  Зеленоградск   
1                  Hotel           Уфа   
2                  Hotel           Уфа   
3                  Hotel         Сибай   
4             Mini-hotel   Стерлитамак   
...                  ...           ...   
5253              Hostel           Уфа   
5254           Apartment           Уфа   
5255               Hotel           Уфа   
5256               Hotel           Уфа   
5257               Hotel           Уфа   

                                                address  rating  rating_5  \
0                 улица Приморская, д. 31, Зеленоградск    10.0      5.00   
1                            улица Мушникова, д.28, Уфа     8.7      4.35   
2                      Коммунистическая ул., д. 53, Уфа     6.3      3.15   
3                           Учалинская улица, 37, Сибай    10.0      5.00   
4                     улица Дружбы, д.28 Д, Стерлитамак     8.0      4.00   
...                                                 ...     ...       ...   
5253  Ulitsa Lenina, 9/11, 3, flat 24, Ufa, Russia, Уфа     7.5      3.75   
5254     Ulitsa Mingazheva 102, 450078 Ufa, Russia, Уфа     0.0      0.00   
5255                                ул. Комарова 1, Уфа     0.0      0.00   
5256                      ул. Менделеева, д. 160/3, Уфа     0.0      0.00   
5257        Poselok Bulgakova, Mikroraion Aeroport, Уфа     0.0      0.00   

      review_count  region  star_rating  rooms_count       uid  \
0                0       2          NaN          NaN  10595357   
1                5       2          NaN          8.0   8332081   
2               10       2          4.0         15.0   7724907   
3                2       2          NaN         15.0   6598197   
4                0       2          2.0         23.0   8622634   
...            ...     ...          ...          ...       ...   
5253             0       2          NaN          NaN   8661195   
5254             0       2          NaN          1.0   6297197   
5255             0       2          NaN          NaN   8753471   
5256            21       2          3.0         18.0   7887775   
5257             0       2          NaN        123.0   6558828   

                   parsing_time        lat        lon  
0    2023-09-26 20:56:02.013106  54.951103  20.456526  
1    2023-09-26 20:56:02.013106  54.783913  56.116150  
2    2023-09-26 20:56:02.013106  54.726463  55.946445  
3    2023-09-26 20:56:02.013106  52.719230  58.658596  
4    2023-09-26 20:56:02.013106  53.642567  55.932686  
...                         ...        ...        ...  
5253 2023-09-26 20:56:02.013106  54.727528  55.947640  
5254 2023-09-26 20:56:02.013106  54.727356  55.968070  
5255 2023-09-26 20:56:02.013106  54.805300  56.036000  
5256 2023-09-26 20:56:02.013106  54.737904  56.029900  
5257 2023-09-26 20:56:02.013106  54.566658  55.886185  

[5258 rows x 16 columns]

## 5.2. Топ 10 по platform1, где больше всего отзывов

In [16]:
top_review = res_data1[['title', 'review_count']].copy().sort_values(by='review_count', ascending=False)
top_review.head(10)

title  review_count
1641        Radisson Blu Hotel Kaliningrad          1201
8                    AZIMUT Сити Отель Уфа           988
1719           Отель Кайзерхоф (Kaiserhof)           930
5222          Отель Ibis Калининград Центр           837
5          Hilton Garden Inn Ufa Riverside           616
11                  Отель АМАКС Сити-отель           540
6              Отель Hampton by Hilton Уфа           460
181   Гостиничный Комплекс Президент Отель           414
5219    Гостиница Мартон Палас Калининград           334
1494                Гостиница "Шкиперская"           302

#### ВЫВОД: больше всего отзывов получила гостиница "Radisson Blu Hotel Kaliningrad" - 1201 отзыв

## 5.3. Квадрат координат размером 1км на 1км, где больше всего гостиниц

### 5.3.1. Определим минимальное и максимальное значение координат:

In [17]:
min_lat = res_data1['lat'].min()
max_lat = res_data1['lat'].max()

min_lon = res_data1['lon'].min()
max_lon = res_data1['lon'].max()

### 5.3.2. Разделим диапазон координат на квадраты размером 1 км:

In [18]:
step = 1 / 111
sq_lat = pd.cut(res_data1['lat'], bins=int((max_lat-min_lat)/step), include_lowest=True)
sq_lon = pd.cut(res_data1['lon'], bins=int((max_lon-min_lon)/step), include_lowest=True)

### 5.3.3. Сгруппируем датасет по комбинации квадратов координат и посчитаем количество гостиниц в каждой группе:

In [19]:
grouped = res_data1.groupby([sq_lat, sq_lon]).size().reset_index(name='количество_гостиниц')

### 5.3.4. Квадрат координат с максимальным количеством гостиниц

In [20]:
max_count_square = grouped[grouped['количество_гостиниц'] == grouped['количество_гостиниц'].max()]

In [21]:
print('Координаты квадрата с наибольшим количеством гостиниц:')
print('lat:', max_count_square['lat'].values[0])
print('lot:', max_count_square['lon'].values[0])

Координаты квадрата с наибольшим количеством гостиниц:
lat: (54.709, 54.718]
lot: (20.499, 20.508]


## 5.4. Карта

In [22]:
res_data_sq1 = res_data1.query('(lat >= 54.709 and lat <= 54.718) and (lon >= 20.499 and lon <= 20.508)')

shir = 54.714
dol = 20.504

karta = folium.Map(location=[shir, dol], zoom_start=5)

def create(row):
    Marker([row['lat'], row['lon']], popup=f"{row['title']}").add_to(karta)
    
res_data_sq1.apply(create, axis=1)

karta

# 6. Предобработка данных data2

In [23]:
data2.head()

id                 create_time                  title  \
0   4608  2023-09-14 15:46:07.855977         Уют, гостиница   
1   4815  2023-09-14 15:46:48.332492      Гостиница Ашкадар   
2   4750  2023-09-14 15:46:35.659931  Островок, база отдыха   
3   4786  2023-09-14 15:46:42.672665     Ардан, база отдыха   
4  16310  2023-09-15 11:56:48.785266              Гостиница   

  hotel_type_original         city  \
0               hotel          NaN   
1                 NaN  sterlitamak   
2          hotel_rest          NaN   
3          hotel_rest          NaN   
4               hotel    beloretsk   

                                             address  rating  rating_5  \
0  Россия, Республика Башкортостан, Аскинский рай...     NaN       NaN   
1  Россия, Республика Башкортостан, Стерлитамак г...     NaN       NaN   
2  Россия, Республика Башкортостан, Гафурийский р...     NaN       NaN   
3  Россия, Республика Башкортостан, Баймакский ра...     NaN       NaN   
4  Россия, Республика Башкортостан, Белорецкий ра...     NaN       NaN   

   review_count  region  star_rating  rooms_count contact_social description  \
0           NaN       2          NaN          5.0            NaN         NaN   
1           NaN       2          NaN          NaN            NaN         NaN   
2           NaN       2          NaN          NaN            NaN         NaN   
3           NaN       2          NaN          NaN            NaN         NaN   
4           NaN       2          NaN          NaN            NaN         NaN   

  email phone website                uid                parsing_time  \
0   NaN   NaN     NaN  70000001064499291  2023-09-14 22:44:57.013238   
1   NaN   NaN     NaN   7600610350202940  2023-09-14 22:44:57.054250   
2   NaN   NaN     NaN  70000001076845858  2023-09-14 22:44:57.043234   
3   NaN   NaN     NaN  70000001075946655  2023-09-14 22:44:57.049244   
4   NaN   NaN     NaN  70000001031030911  2023-09-15 18:55:23.191778   

         lat        lon  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4  53.969912  58.400557

In [24]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2767 entries, 0 to 2766
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   2767 non-null   int64  
 1   create_time          2767 non-null   object 
 2   title                2767 non-null   object 
 3   hotel_type_original  2747 non-null   object 
 4   city                 1622 non-null   object 
 5   address              2767 non-null   object 
 6   rating               2119 non-null   float64
 7   rating_5             2119 non-null   float64
 8   review_count         2119 non-null   float64
 9   region               2767 non-null   int64  
 10  star_rating          591 non-null    float64
 11  rooms_count          1662 non-null   float64
 12  contact_social       506 non-null    object 
 13  description          197 non-null    object 
 14  email                67 non-null     object 
 15  phone                1076 non-null   o

#### Массив данных содержит 21 колонку и 2767 строки

## 6.1. Пропуски

#### Заменить пропуски в колонках не представляется возможным, так как данные о гостиницах реальны и замена пустых значений средним значением или медианой не будет отражать реальных данных гостиниц.

## 6.2. Типы данных

#### Содержание всех колонок, кроме 'create_time' и 'parsing_time' соответствует типу данных в колонке. Исправим тип данных в колонках о времени

In [25]:
data2['create_time'] = pd.to_datetime(data1['create_time'])

In [26]:
data2['parsing_time'] = pd.to_datetime(data1['parsing_time'])

In [27]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2767 entries, 0 to 2766
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   2767 non-null   int64         
 1   create_time          2767 non-null   datetime64[ns]
 2   title                2767 non-null   object        
 3   hotel_type_original  2747 non-null   object        
 4   city                 1622 non-null   object        
 5   address              2767 non-null   object        
 6   rating               2119 non-null   float64       
 7   rating_5             2119 non-null   float64       
 8   review_count         2119 non-null   float64       
 9   region               2767 non-null   int64         
 10  star_rating          591 non-null    float64       
 11  rooms_count          1662 non-null   float64       
 12  contact_social       506 non-null    object        
 13  description          197 non-null

#### ВЫВОД: тип данных в колонках со временем исправлен

## 6.3. Дубликаты

In [28]:
data2.duplicated().sum()

0

#### ВЫВОД: дубликатов не обнаружено

In [29]:
#data2.drop('contact_social', axis=1, inplace=True)
#data2.drop('description', axis=1, inplace=True)
#data2.drop('email', axis=1, inplace=True)
#data2.drop('phone', axis=1, inplace=True)
#data2.drop('website', axis=1, inplace=True)

# 7. Исследовательский анализ данных

In [30]:
title2 = data2['title'].value_counts().reset_index()
title2 = title2.query('count > 1')
title2

title  count
0                                    Гостиница     65
1                               Уют, гостиница     60
2                                   Мини-отель     20
3                               Автобус в лесу     20
4                           Евразия, гостиница     20
..                                         ...    ...
478                      Людмила, гостевой дом      5
479              В гостях у дворика, гостиница      5
480                         100 друзей, хостел      5
481  Урал, гостинично-оздоровительный комплекс      5
482                          Лена, база отдыха      5

[483 rows x 2 columns]

## 7.1. Актуальные данные по гостиницам

#### Финальный список гостиниц, в котором по каждой гостинице - только одна запись с самыми актуальными и полными данными

In [31]:
agg_data2 = data2.groupby('title')['create_time'].max().reset_index()

res_data2 = pd.merge(data2, agg_data2, on=['title', 'create_time'], how='inner')
res_data2

id                create_time                       title  \
0     4668 2023-09-26 15:01:49.233151                 Три пескаря   
1     4700 2023-09-26 15:01:54.124427        Белоречье, гостиница   
2     4808 2023-09-26 15:02:16.705051  Любимое место, база отдыха   
3    16312 2023-09-26 15:02:17.745185     Дальнобойщик, гостиница   
4    16313 2023-09-26 15:02:17.956962           Родник, гостиница   
..     ...                        ...                         ...   
478  67697 2023-09-26 15:05:08.947118           Азалия, гостиница   
479  67698 2023-09-26 15:05:09.159092            Инзер, гостиница   
480  67699 2023-09-26 15:05:09.371139        By Richotels, хостел   
481  67701 2023-09-26 15:05:09.583508              Уют, гостиница   
482  67702 2023-09-26 15:05:09.795208            Замок, гостиница   

    hotel_type_original       city  \
0            hotel_rest        NaN   
1                 hotel  beloretsk   
2            hotel_rest        NaN   
3                 hotel        NaN   
4                 hotel        NaN   
..                  ...        ...   
478               hotel        ufa   
479               hotel        NaN   
480               hotel        ufa   
481               hotel        NaN   
482               hotel        NaN   

                                               address  rating  rating_5  \
0    Россия, Республика Башкортостан, Караидельский...     4.7       4.7   
1    Россия, Республика Башкортостан, Белорецкий ра...     NaN       NaN   
2    Россия, Республика Башкортостан, Караидельский...     NaN       NaN   
3    Россия, Республика Башкортостан, Белорецкий ра...     1.0       1.0   
4    Россия, Республика Башкортостан, Кугарчинский ...     NaN       NaN   
..                                                 ...     ...       ...   
478  Россия, Республика Башкортостан, Уфа городской...     2.5       2.5   
479  Россия, Республика Башкортостан, Белорецкий ра...     1.4       1.4   
480  Россия, Республика Башкортостан, Уфа городской...     4.5       4.5   
481  Россия, Республика Башкортостан, Хайбуллинский...     2.3       2.3   
482  Россия, Республика Башкортостан, Балтачевский ...     1.1       1.1   

     review_count  region  star_rating  rooms_count  \
0             3.0       2          NaN          NaN   
1             NaN       2          NaN          NaN   
2             NaN       2          NaN          NaN   
3             1.0       2          NaN          5.0   
4             NaN       2          NaN          6.0   
..            ...     ...          ...          ...   
478           4.0       2          NaN          3.0   
479          11.0       2          NaN          4.0   
480         217.0       2          NaN          NaN   
481           3.0       2          NaN          9.0   
482           6.0       2          NaN          NaN   

                                        contact_social  \
0                                                  NaN   
1                                                  NaN   
2                                                  NaN   
3                                                  NaN   
4                                                  NaN   
..                                                 ...   
478                                                NaN   
479                                                NaN   
480  [{"social": "you_tube", "contact": "https://yo...   
481                                                NaN   
482                                                NaN   

                                           description  \
0                                                  NaN   
1                                                  NaN   
2                                                  NaN   
3                                                  NaN   
4                                                  NaN   
..                                                 ...   
478                                          

## 7.2. Топ 10 по platform2, где больше всего отзывов

In [32]:
top_review2 = res_data2[['title', 'review_count']].copy().sort_values(by='review_count', ascending=False)
top_review2.head(10)

title  review_count
277       Hilton Garden Inn Ufa Riverside, отель         823.0
274               Башкирия, гостиничный комплекс         780.0
282                 Hampton by hilton Ufa, отель         451.0
281               Урал тау, гостиничный комплекс         416.0
278                        AZIMUT Сити Отель Уфа         386.0
284                  Амакс Сити-отель, гостиница         358.0
276        Президент отель, гостиничный комплекс         341.0
275  Sheratonplaza Ufa Congress Hotel, гостиница         240.0
297                             ВолнаСити, отель         232.0
264                  Свояк, гостиничный комплекс         222.0

#### ВЫВОД: больше всего отзывов получила гостиница Hilton Garden Inn Ufa Riverside, отель - 823 отзывов

## 7.3. Квадрат координат размером 1км на 1км, где больше всего гостиниц

In [33]:
min_lat2 = res_data2['lat'].min()
max_lat2 = res_data2['lat'].max()

min_lon2 = res_data2['lon'].min()
max_lon2 = res_data2['lon'].max()

In [34]:
step2 = 1 / 111
sq_lat2 = pd.cut(res_data2['lat'], bins=int((max_lat2-min_lat2)/step2), include_lowest=True)
sq_lon2 = pd.cut(res_data2['lon'], bins=int((max_lon2-min_lon2)/step2), include_lowest=True)

In [35]:
grouped2 = res_data2.groupby([sq_lat2, sq_lon2]).size().reset_index(name='количество_гостиниц')

In [36]:
max_count_square2 = grouped2[grouped2['количество_гостиниц'] == grouped2['количество_гостиниц'].max()]

In [37]:
print('Координаты квадрата с наибольшим количеством гостиниц:')
print('lat:', max_count_square2['lat'].values[0])
print('lot:', max_count_square2['lon'].values[0])

Координаты квадрата с наибольшим количеством гостиниц:
lat: (53.806, 53.815]
lot: (58.633, 58.642]


## 7.4. Карта

In [38]:
res_data_sq2 = res_data2.query('(lat >= 53.806 and lat <= 53.815) and (lon >= 58.633 and lon <= 58.642)')

shir2 = 53.810
dol2 = 58.638

karta2 = folium.Map(location=[shir2, dol2], zoom_start=5)

def create(row):
    Marker([row['lat'], row['lon']], popup=f"{row['title']}").add_to(karta2)
    
res_data_sq2.apply(create, axis=1)

karta2

# ВЫВОД: В процессе работы я предобработал датасеты и нашёл ответы на поставленные вопросы